### PR TITLE
Evaluate performance of covariates at predicting various mutations

### PR DESCRIPTION
**Pull request by dhimmel**
Creates an explore directory and README for this type of exploratory notebook.

See how well covariates (non-expression features) predict TP53 mutation.

Related to https://github.com/cognoma/machine-learning/issues/8: General mutation-load does provide some ability to predict mutation status of TP53.

Partially addresses https://github.com/cognoma/machine-learning/issues/21: Covariates are extracted from `samples.tsv`.

### CODE REVIEW

**Comment by cgreene** [comment]
> Related to #8: General mutation-load does provide some ability to predict mutation status of TP53.

This isn't too surprising given TP53's role in controlling cell cycle checkpoints. Is this only true for genes in cell cycle checkpoint or DNA repair pathways, or is it also true for genes in proliferation pathways?

**Comment by dhimmel** [comment]
> Is this only true for genes in cell cycle checkpoint or DNA repair pathways, or is it also true for genes in proliferation pathways?

@cgreene, I updated the notebook to evaluate performance of covariate-only classifiers for the 8 interesting mutations we've [previously considered](https://github.com/cognoma/cancer-data/pull/22#issuecomment-245294155). Here is performance of the models with all covariates included:

covariate-performance

So actually, _TP53_ is among the hardest to predict using only covariates. _VHL_ which is highly disease-specific achieves a near-perfect AUROC. Therefore, without the disease or organ covariate, expression classifiers of _VHL_ are likely just classifying kidney clear cell carcinoma / kidney tissue.

See [this dataframe](https://github.com/dhimmel/machine-learning/blob/50aa17482dd1f9ebfae7d98fcf98a0209eb86487/explore/confounding/auroc.tsv) to get a general idea of covariate importance. Mutation load and disease type both seem important.

**Comment by cgreene** [comment]
@dhimmel : Interesting! I can imagine that gene expression would clearly capture disease/organ, so if one usually had a mutation in a single gene that was relatively specific (e.g. VHL) I could imagine that creating a strong signal.

File: explore/confounding/confounding.py [hunk 1]
```diff
@@ -0,0 +1,280 @@
+
```
Comment (gwaybio) [inline explore/confounding/confounding.py:1]: looks like you need to `nbconvert` this again

Comment (gwaybio) [reply explore/confounding/confounding.py:1]: I'll take a look again once it's easier to read

**Review by gwaybio — CHANGES_REQUESTED** [review]
looks like nbconvert `confounding.ipynb` has to be rerun

File: explore/confounding/confounding.py (hunk 1 above)
Comment (dhimmel) [reply explore/confounding/confounding.py:1]: Good catch. should be fixed now.

File: explore/README.md [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# A directory for exploratory machine learning analyses
+
+This directory is home is exploratory analyses that help answer questions about how we should do machine learning. For algorithm implementations see the [`algorithms`](../algorithms) directory. For other types of analyses, place them here.
```
Comment (gwaybio) [inline explore/README.md]: "This directory is home **to** exploratory analyses"?

File: explore/confounding/confounding.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+# coding: utf-8
+
+# # Create a logistic regression model to several mutations from covariates
```
Comment (gwaybio) [inline explore/confounding/confounding.py]: "Create a logistic regression model to **predict** several mutations from covariates"?

File: explore/confounding/confounding.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+
+# In[6]:
+
+def expand_grid(data_dict):
+    """Create a dataframe from every combination of given values."""
+    rows = itertools.product(*data_dict.values())
+    return pd.DataFrame.from_records(rows, columns=data_dict.keys())
+
+mutations = {
+    '7157': 'TP53', # tumor protein p53
```
Comment (gwaybio) [inline explore/confounding/confounding.py]: pep8 says inline comments should be separated by [two spaces](https://www.python.org/dev/peps/pep-0008/#inline-comments)

File: explore/confounding/confounding.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+    'survival': ['alive', 'dead'],
+}
+
+
+# ## Compute performance
+
+# In[8]:
+
+def get_aurocs(X, y, series):
+    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=0)
```
Comment (gwaybio) [inline explore/confounding/confounding.py:135]: what does `series` look like?

**Review by gwaybio — CHANGES_REQUESTED** [review]
Interesting and informative analysis. My comments are generally cosmetic, but it would be nice to get some additional documentation on some of the functions!

**Review by gwaybio — APPROVED** [review]

**Pull request merged**